syntax = "proto3";

package user_service;

// User message
message User {
  int32 id = 1;
  string name = 2;
  string email = 3;
}

// Request messages
message GetUsersRequest {
  int32 skip = 1;
  int32 limit = 2;
}

message GetUserRequest {
  int32 user_id = 1;
}

message CreateUserRequest {
  string name = 1;
  string email = 2;
}

message UpdateUserRequest {
  int32 user_id = 1;
  string name = 2;
  string email = 3;
}

message DeleteUserRequest {
  int32 user_id = 1;
}

message SearchUsersRequest {
  string query = 1;
}

// Response messages
message GetUsersResponse {
  repeated User users = 1;
}

message GetUserResponse {
  User user = 1;
  bool found = 2;
}

message CreateUserResponse {
  User user = 1;
  bool success = 2;
}

message UpdateUserResponse {
  User user = 1;
  bool success = 2;
}

message DeleteUserResponse {
  bool success = 1;
  string message = 2;
}

message SearchUsersResponse {
  repeated User users = 1;
}

// User service definition
service UserService {
  rpc GetUsers(GetUsersRequest) returns (GetUsersResponse);
  rpc GetUser(GetUserRequest) returns (GetUserResponse);
  rpc CreateUser(CreateUserRequest) returns (CreateUserResponse);
  rpc UpdateUser(UpdateUserRequest) returns (UpdateUserResponse);
  rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse);
  rpc SearchUsers(SearchUsersRequest) returns (SearchUsersResponse);
}